src/ripple/app/tx/impl/CreateOffer.cpp

```diff
@@ -629,6 +626,419 @@ CreateOffer::cross (
     }
 }
 
+std::pair<TER, Amounts>
+CreateOffer::crossV2 (
+    PaymentSandbox& psb,
+    PaymentSandbox& psbCancel,
+    Amounts const& takerAmount)
+{
+    try
+    {
+        // If the taker is unfunded before we begin crossing there's nothing
+        // to do - just return an error.
+        //
+        // We check this in preclaim, but when selling XRP charged fees can
+        // cause a user's available balance to go to 0 (by causing it to dip
+        // below the reserve) so we check this case again.
+        STAmount const inStartBalance = accountFunds (
+            psb, account_, takerAmount.in, fhZERO_IF_FROZEN, j_);
+        if (inStartBalance.signum() <= 0)
+        {
+            // The account balance can't cover even part of the offer.
+            JLOG (j_.debug()) <<
+                "Not crossing: taker is unfunded.";
+            return { tecUNFUNDED_OFFER, takerAmount };
+        }
+
+        // If the offer is passive don't even attempt to cross.
+        std::uint32_t const txFlags =ctx_.tx.getFlags();
+        if (txFlags & tfPassive)
+        {
+            return { tesSUCCESS, takerAmount };
```

Comment:
This is just a quick comment based on a cursory reading; I haven't delved deeply at the code, but I'm not sure that this is the right thing.

tfPassive means to not consume offers at the same quality level, but crossing still needs to be performed, as there might be offers at a better quality level. This code seems to just not do any crossing when tfPassive is set.

Am I missing something obvious here?